Added support for v4 signing to S3.

kat-co opened the pull request:
- S3 constructor now takes a Signer function. This indicates the preferred signing methodology.
- Users can still use the recommended type of signing by using the Signer defined on the region.

kat-co committed:
Added support for v4 signing to S3.

- S3 constructor now takes a Signer function. This indicates the preferred signing methodology.
- Users can still use the recommended type of signing by using the Signer defined on the region.

kat-co commented:
Most of this PR is switching the S3 requests over to vanilla http.Request structures. This was done because the new AWS signing operates on these.

dimitern commented on aws/aws.go:
```diff
@@ -30,6 +31,13 @@ type Region struct {
+	if r.S3BucketEndpoint != "" {
+		return strings.Replace(r.S3BucketEndpoint, "${bucket}", bucketName, -1)
+	}
+	return r.S3Endpoint + "/" + bucketName + "/"
```
path.Join here instead?

kat-co replied:
path.Join scrubs the content and will remove one of the slashes in "http://" in the r.S3Endpoint.

dimitern commented on s3/multi.go:
```diff
@@ -49,39 +50,50 @@ type listMultiResp struct {
+	req.URL.RawQuery = query.Encode()
+
+	for attempt := attempts.Start(); attempt.Next(); attempt = attempts.Start() {
+
```
d?

kat-co replied:
Present for readability.

dimitern commented on s3/multi.go:
```diff
@@ -49,39 +50,50 @@ type listMultiResp struct {
+		query.Set("upload-id-marker", multiResp.NextUploadIdMarker)
+		req.URL.RawQuery = query.Encode()
 	}
+
```
d

kat-co replied:
Present for readability.

dimitern commented on s3/multi.go:
```diff
@@ -106,35 +118,37 @@ func (b *Bucket) Multi(key, contType string, perm ACL) (*Multi, error) {
+	query := req.URL.Query()
+	query.Add("upload", "")
+	req.URL.RawQuery = query.Encode()
+
```
d?

kat-co replied:
Present for readability.

dimitern commented on s3/s3i_test.go:
```diff
@@ -239,9 +208,10 @@ func (s *ClientTests) TestRegions(c *C) {
-			_, err := b.Get("non-existent")
+			s := s3.New(s.s3.Auth, r, aws.SignV2)
+			b, err := s.Bucket("goamz-" + s.Auth.AccessKey)
+			c.Assert(err, IsNil)
```
I'd prefer not to use c.Assert in a loop like this, so that a single failure won't stop the test. How about:

``` go
if !c.Check(err, IsNil) {
  errs <- err
}
_, err = b.Get("non-existent")
```

kat-co replied:
Good suggestion, I've implemented as you have it laid out.

dimitern commented on s3/s3_test.go:
```diff
@@ -28,8 +28,11 @@ var testServer = testutil.NewHTTPServer()
-	s.s3 = s3.New(auth, aws.Region{Name: "faux-region-1", S3Endpoint: testServer.URL, Sign: aws.SignV2})
+	s.s3 = s3.New(
+		aws.Auth{"abc", "123"},
+		aws.Region{Name: "faux-region-1", S3Endpoint: testServer.URL, Sign: aws.SignV2},
```
Split aws.Region fields in multiple lines, as it's getting a bit too long.

kat-co replied:
Done

dimitern commented on s3/s3.go:
```diff
@@ -577,3 +438,23 @@ func hasCode(err error, code string) bool {
+		return resp, nil
+	}
+
+	panic("could not complete the request within the specified retry attempts")
```
Why not just return an error here? This is production code and I'd rather handle the error outside than to crash the whole process.

kat-co replied:
This was the pattern established when I began working on the code, but you're correct. I've switched it to an error.

dimitern commented on s3/s3.go:
```diff
@@ -309,69 +318,33 @@ type Key struct {
-}
+	req.Close = true
 
-// URL returns a non-signed URL that allows retriving the
```
So we're dropping URL and SignedURL methods entirely? This is not backwards-compatible and will break existing client. Is it possible to leave these two methods in place and somehow simulate what they do with the new logic?

kat-co replied:
As we discussed:
- This is proposed against v3-unstable, so it will not break anyone.
- SignedURL no longer makes sense as an option since v4 utilizes headers for signing.

I have added the URL function back in.

dimitern commented:
Most of this looks good, however dropping the 2 methods I can't quite get. Also, have you tried running all the s3 tests with -amazon to make sure both the s3test server still works (I'm assuming tests pass, otherwise travis CI would've reported it) and the changes work against live AWS and s3test the same way?

dimitern commented:
Sorry, I've seen this is actually targeted to v3-unstable, so changes to the interface is fine at this point. However, I still think URL is worth keeping as a method. I'd like to have another look when you're done with the tests.

kat-co committed:
v4 Signing tweaks & API fixes.

dimitern commented on aws/sign.go:
```diff
@@ -224,37 +239,50 @@ func canonicalQueryString(queryVals url.Values) (string, error) {
+	if req.Body != nil {
+		if b, err := ioutil.ReadAll(req.Body); err != nil {
+			return "", err
+		} else {
```
I'd pull b, err := ioutil.ReadAll(..) before the if and drop the else clause to reduce the indentation level.

dimitern commented on aws/sign.go:
```diff
@@ -224,37 +239,50 @@ func canonicalQueryString(queryVals url.Values) (string, error) {
+			req.Body = ioutil.NopCloser(bytes.NewBuffer(b))
+			return hasher(b), nil
+		}
 	} else {
```
Drop the else and just  `return hasher([]byte("")), nil`

dimitern commented on aws/sign.go:
```diff
@@ -224,37 +239,50 @@ func canonicalQueryString(queryVals url.Values) (string, error) {
-		sort.Strings(sortedHdrVals)
-		hdrVals := strings.Join(sortedHdrVals, ",")
+
+		var hdrVals string
```
How about:

``` go
hdrVals := host
if hName != "host" {
    canonHdrKey := http.CanonicalHeaderKey(hName)
    sortedHdrVals := hdr[canonHdrKey]
    sort.Strings(sortedHdrVals)
    hdrVals = strings.Join(sortedHdrVals, ",")
}
```

dimitern commented on aws/sign_test.go:
```diff
@@ -3,15 +3,18 @@ package aws
 
 type SigningSuite struct{}
 
+const v4skipReason = `"The signing methodology is a "one size fits all" approach. The hashes we check against don't include headers that are added in as requisite parts for S3. That doesn't mean the tests are invalid, or that signing is broken for these examples, but as long as we're adding heads in, it's impossible to know what the new signature should be. We should revaluate these later.`
```
How about:

``` go
// TODO(katco-): The signing methodology is a "one size fits all" approach.
// The hashes we check against against don't include headers that are added
// in as requisite parts for S3. That doesn't mean the tests are invalid, or that
// signing is broken for these examples, but as long as we're adding heads in,
// it's impossible to know what the new signature should be. We should reevaluate
// these later. See issue #XYZ.
const v4SkipReason = `Extra headers present - cannot predict generated signature (issue #XYZ)`
```

This will show nicely as we're running tests (i.e. not splitting the output on multiple lines), but also gives the reader of this source enough context to understand the issue. Please, file an issue for this and replace the #XYZ with the issue number.

dimitern replied:
As for the impossibility of knowing what the hash could be, I guess this is because the hash changes because of the current timestamp (or reqTime). This could be faked for the test servers - setting a fixed time of a request/response. But that's a future improvement. As long as you've actually manually tested the scenarios in the skipped tests I think it's fine.

dimitern commented on exp/sdb/sdb_test.go:
```diff
@@ -25,7 +25,7 @@ var testServer = testutil.NewHTTPServer()
 func (s *S) SetUpSuite(c *C) {
 	testServer.Start()
 	auth := aws.Auth{"abc", "123"}
-	s.sdb = sdb.New(auth, aws.Region{SDBEndpoint: testServer.URL, Sign: aws.SignV2})
```
It's great you took the time to actually fix the exp/ packages as well. However, they're barely working and are not maintained, so if there are test failures here I don't really care.

dimitern commented on s3/multi_test.go:
```diff
@@ -160,17 +165,18 @@ func (s *S) TestPutAllNoPreviousUpload(c *C) {
 	c.Assert(err, IsNil)
 
 	parts, err := multi.PutAll(strings.NewReader("part1part2last"), 5)
-	c.Assert(parts, HasLen, 3)
```
Good catch!

dimitern commented on s3/s3.go:
```diff
@@ -151,78 +167,98 @@ func (b *Bucket) Get(path string) (data []byte, err error) {
+	return resp.Body, nil
 }
 
 // Put inserts an object into the S3 bucket.
```
s/Put/PutReader/

dimitern replied:
Or actually, just drop the Put doc comment and body only. rather than PutReader's doc comment as well.

dimitern commented on s3/s3.go:
```diff
@@ -151,78 +167,98 @@ func (b *Bucket) Get(path string) (data []byte, err error) {
 	return b.PutReaderWithHeader(path, r, length, contType, perm, http.Header{})
 }
 
 // PutReader inserts an object into the S3 bucket by consuming data
```
s/PutReader/PutReaderWithHeader/

dimitern replied:
Also, restore the two lines below in the doc comment - ... The HTTP PUT request will have...

dimitern commented on s3/s3_test.go:
```diff
@@ -96,27 +104,11 @@ func (s *S) TestGet(c *C) {
 	c.Assert(string(data), Equals, "content")
 }
 
-func (s *S) TestURL(c *C) {
```
Since URL() is back, should this test stay as well?

dimitern commented:
All in all it looks fine, but I'm getting some s3 test failures in both local and live tests. I've sent you a mail with more details.

dimitern commented:
LGTM with the patch discussed on IRC: http://pastebin.ubuntu.com/10490009/ which makes s3 live tests to pass reliably. However, a side-effect of that is leaving a bunch of buckets after each live test run - please, add an issue for this as well. When we have time, we'll fix it.

kat-co committed:
Addressed review issues.

- Minor formatting issues.
- Changed skip message to a comment. Now referencing bug#.
- Documentation cleaned up.

dimitern commented:
Great, I'm merging this.
